src/main/java/org/mule/extension/sftp/internal/operation/AbstractFileCommand.java

```diff
@@ -0,0 +1,280 @@
+   */
+  protected I resolvePath(String filePath) {
+    I path = getBasePath(fileSystem);
+    if (filePath != null) {
+      path = resolvePath(path, filePath);
+    }
+    return getAbsolutePath(path);
+  }
+
+  /**
+   * @param fileName the name of a file
+   * @return {@code true} if {@code fileName} equals to &quot;.&quot; or &quot;..&quot;
+   */
+  protected boolean isVirtualDirectory(String fileName) {
+    return ".".equals(fileName) || "..".equals(fileName);
+  }
+
+  /**
+   * Similar to {@link #resolvePath(String)} only that it throws a {@link IllegalArgumentException} if the given path doesn't
+   * exist.
+   * <p>
+   * The existence of the obtained path is verified by delegating into {@link #exists(I)}
+   *
+   * @param filePath the path to a file or directory
+   * @return an absolute path
+   */
+  protected I resolveExistingPath(String filePath) {
+    I path = resolvePath(filePath);
+    if (!exists(path)) {
+      throw pathNotFoundException(path);
+    }
+
+    return path;
+  }
+
+  /**
+   * Returns an {@link IllegalPathException} explaining that a {@link FileSystem#read(FileConnectorConfig, String, boolean)}
+   * operation was attempted on a {@code path} pointing to a directory
+   *
+   * @param path the path on which a read was attempted
+   * @return {@link IllegalPathException}
+   */
+  protected IllegalPathException cannotReadDirectoryException(I path) {
+    return new IllegalPathException(format("Cannot read path '%s' since it's a directory", pathToString(path)));
+  }
+
+  /**
+   * Returns a {@link IllegalPathException} explaining that a
+   * {@link FileSystem#list(FileConnectorConfig, String, boolean, Predicate)} operation was attempted on a {@code path} pointing
+   * to a file.
+   *
+   * @param path the path on which a list was attempted
+   * @return {@link IllegalPathException}
+   */
+  protected IllegalPathException cannotListFileException(I path) {
+    return new IllegalPathException(format("Cannot list path '%s' because it's a file. Only directories can be listed",
+                                           pathToString(path)));
+  }
+
+  /**
+   * Returns a {@link IllegalPathException} explaining that a
+   * {@link FileSystem#list(FileConnectorConfig, String, boolean, Predicate)} operation was attempted on a {@code path} pointing
+   * to a file.
+   *
+   * @param path the on which a list was attempted
+   * @return {@link RuntimeException}
+   */
+  protected IllegalPathException pathNotFoundException(I path) {
+    return new IllegalPathException(format("Path '%s' doesn't exist", pathToString(path)));
+  }
+
+  /**
+   * Returns a {@link IllegalPathException} explaining that an operation is trying to write to the given {@code path} but it
+   * already exists and no overwrite instruction was provided.
+   *
+   * @param path the that the operation tried to modify
+   * @return {@link IllegalPathException}
+   */
+  public FileAlreadyExistsException alreadyExistsException(I path) {
```

Comment:
Why is this public? All the neighboring exception methods are protected, I assume, so that the implementing classes can throw the exceptions in a consistent way. All the accesses to this method seem coming from classes in the inheritance hierarchy.